### PR TITLE
Milestone: set up associations :arrows_counterclockwise:

### PR DESCRIPTION
**Pull request by ahmedeid6842**
🚀 **Project Requirements: Classroom, Book, Rental, and Student Classes**

This pull request adds the necessary classes and associations to fulfill the project requirements. The changes include the following:

## Done: :white_check_mark: 
1. Created the `Classroom` class with the following features:
   - [x] Initialized the `@label` instance variable in the constructor.
   - [x] Added a setter and getter for `@label` using `attr_accessor`.
   - [x] Implemented a has-many/belongs-to relationship between `Classroom` and `Student`.
   - [x] Ensured that adding a student to a classroom also sets the classroom for the student, and vice versa.

2. Added the `Book` class with the following features:
   - [x] Initialized the `@title` and `@author` instance variables in the constructor.
   - [x] Added setters and getters for the instance variables using `attr_accessor`.
   - [x] Included a `@rentals` array to establish the has-many relationship between `Book` and `Rental`.

3. Implemented the `Rental` class with the following features:
   - [x] Initialized the `@date` instance variable in the constructor.
   - [x] Added setters and getters for `@date` using `attr_accessor`.
   - [x] Established the many-to-many relationship between `Person` and `Book` using the `Rental` class.
   - [x] Created the has-many side for `Book` and `Rental`, and the belongs-to side for `Rental` and `Book`.
   - [x] Created the has-many side for `Person` and `Rental`, and the belongs-to side for `Rental` and `Person`.
   - [x] Modified the constructor of `Rental` to set the associated `Book` and `Person` instances.

4. Added the `Student` class as a subclass of `Person` with the following features:
   - [x] Included an `attr_accessor` for `classroom`.
   - [x] Updated the constructor to pass the parameters correctly to the superclass `Person`.
   - [x] Implemented the `play_hockey` method as an example.

## 💡 **Highlights:**

- Implemented the required classes `Classroom`, `Book`, `Rental`, and `Student` with the necessary attributes and associations.
- Established the has-many and belongs-to relationships between `Classroom` and `Student`, and the many-to-many relationship between `Person` and `Book` using the intermediate class `Rental`.
- Ensured that adding a student to a classroom sets the classroom for the student, and vice versa.
- Added appropriate getters and setters for instance variables using `attr_accessor`.

🎉 This contribution fulfills the project requirements and enhances the codebase with the necessary classes and associations. 
## Thanks in advance for taking time and effort to review my code :pray: 

giphy

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -3,9 +3,10 @@
 class Student < Person
   attr_accessor :classroom
 
-  def initialize(id, age, classroom, name = 'Unknown', parent_permission: true)
-    super(id, age, name, parent_permission)
+  def initialize(age, classroom, name = 'Unknown', parent_permission: true)
+    super(age, name, parent_permission)
```
Comment (tresorsawasawa) [inline student.rb:7]: - [ ] To correctly complete the `has-many / belongs-to relationships` between `Clasroom` and `Student` classes, kindly ensure that the `Student` class has a method that serves as the `belongs-to` side (a student belongs to a classroom). It should look to something like this:
```ruby
class Student < Person
  attr_reader :classroom
  .
  .
  def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
+
```
Comment (tresorsawasawa) [inline book.rb:4]: - [ ] To correctly complete the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` classes using the intermediate class `Rental`, kindly ensure that `Getter` and `Setters` for `@rentals` sre present i.e. `attr_accessor :title, :author, :rentals`

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+end
```
Comment (tresorsawasawa) [inline book.rb:10]: - [ ] To correctly complete the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` classes using the intermediate class `Rental`, kindly ensure that a method for adding rentals is present. It should look similar to this:
```ruby
class Book
   .
   .
   .
  def add_rental(person, date)
    # *the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. 
    Rental.new(date, self, person)
  end
end
```

File: person.rb [hunk 4]
```diff
@@ -5,14 +5,15 @@
...
   def initialize(age, name = 'Unknown', parent_permission: true)
     super()
     @id = rand(1..1000)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rentals = []
   end
 
   def correct_name
```
Comment (tresorsawasawa) [inline person.rb:19]: - [ ] As person has many **rentals**, kindly ensure that a method for adding rentals is present present in your `Pesron` class. The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association. It should look similar to this:
```ruby
  class Person < Nameable
  .
  .
  . 
  def add_rental(book, date)
    Rental.new(date, book, self)
  end
  .
  .
end
```

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @ahmedeid6842 
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

almost-there

## To highlight :champagne: 
- Follow Gitflow :heavy_check_mark: 
- Most of relationship are respected :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by DuaneDave — APPROVED** [review]
Hi @ahmedeid6842 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✔️ All associations are correctly implemented

## Optional suggestions
N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**